program/src/processor/initialize_sol_staker_stake.rs

```diff
@@ -0,0 +1,187 @@
+use paladin_sol_stake_view_program_client::{
+    instructions::GetStakeActivatingAndDeactivatingCpiBuilder,
+    GetStakeActivatingAndDeactivatingReturnData,
+};
+use solana_program::{
+    entrypoint::ProgramResult,
+    program::{get_return_data, invoke_signed},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    system_instruction,
+};
+
+use crate::{
+    err,
+    error::StakeError,
+    instruction::accounts::{Context, InitializeSolStakerStakeAccounts},
+    processor::unpack_initialized_mut,
+    require,
+    state::{
+        find_sol_staker_stake_pda, find_validator_stake_pda, get_sol_staker_stake_pda_signer_seeds,
+        Config, SolStakerStake, ValidatorStake,
+    },
+};
+
+/// Initializes stake account data for a SOL staker.
+///
+/// There can be only one SOL staker stake account per stake state and config account, since
+/// the stake state is part of the SOL staker stake account seeds.
+///
+/// NOTE: Anybody can create the stake account for a SOL staker. For new
+/// accounts, the authority is initialized to the stake state account's withdrawer.
+///
+/// 0. `[]` Stake config account
+/// 1. `[w]` SOL staker stake account
+///     * PDA seeds: ['stake::state::sol_staker_stake', stake state, config]
+/// 2. `[]` SOL stake state account
+/// 3. `[]` Stake history sysvar
+/// 4. `[]` System program
+/// 5. `[]` Paladin SOL Stake View program
+#[allow(clippy::useless_conversion)]
+pub fn process_initialize_sol_staker_stake(
+    program_id: &Pubkey,
+    ctx: Context<InitializeSolStakerStakeAccounts>,
+) -> ProgramResult {
+    // Accounts validation.
+
+    // config
+    // - owner must the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let data = &ctx.accounts.config.try_borrow_data()?;
+
+    let config = bytemuck::try_from_bytes::<Config>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config"
+    );
+
+    // stake state (validated by the SOL Stake View program)
+    // - must have a delegation
+
+    GetStakeActivatingAndDeactivatingCpiBuilder::new(ctx.accounts.sol_stake_view_program)
+        .stake(ctx.accounts.stake_state)
+        .stake_history(ctx.accounts.stake_history)
+        .invoke()?;
```

Comment:
Does this validate somewhere that the `sol_stake_view_program` is correct? That program is a huge attack target, so we should definitely be sure that we're calling the correct one!

Reply:
Yes, it is indirectly validated since the CPI instruction is built with the address of the correct program. This is the error that you get when you pass the wrong program.
```bash
DEBUG Instruction: InitializeSolStakerStake
DEBUG Unknown program stkVUdWUiarMmkttUKMGLCLwHUkBqYfQ9vZcfG3T7LU
DEBUG Program PStake1111111111111111111111111111111111111 consumed 3037 of 399850 compute units
DEBUG Program PStake1111111111111111111111111111111111111 failed: An account required by the instruction is missing
```

But I will add a test so the error message is more clear.

Reply:
Done!
```
DEBUG Instruction: InitializeSolStakerStake
DEBUG Constraint failed: invalid sol stake view program
DEBUG Error: IncorrectProgramId
DEBUG Program PStake1111111111111111111111111111111111111 consumed 1778 of 399850 compute units
DEBUG Program PStake1111111111111111111111111111111111111 failed: incorrect program id for instruction
```